File: 백준/Gold/1914. 하노이 탑/하노이 탑.py
import sys

N = int(sys.stdin.readline().strip())

def hanoi_tower(N, move_from, move_to, move_through):
    if N == 1:
        print(move_from, move_to)
        return
    hanoi_tower(N - 1, move_from, move_through, move_to)
    print(move_from, move_to)
    hanoi_tower(N - 1, move_through, move_to, move_from)

print(2**N - 1)

if N <= 20:
    hanoi_tower(N, 1, 3, 2)

File: 백준/Gold/1914. 하노이 탑/하노이 탑.java
import java.math.BigInteger;
import java.util.Scanner;

public class Main {
    private static void hanoiTower(int n, int moveFrom, int moveTo, int moveThrough){
        if (n == 1){
            System.out.println(moveFrom + " " + moveTo);
            return;
        }

        hanoiTower(n - 1, moveFrom, moveThrough, moveTo);
        System.out.println(moveFrom + " " + moveTo);
        hanoiTower(n - 1, moveThrough, moveTo, moveFrom);
    }

    public static void main(String[] args){
        Scanner sc = new Scanner(System.in);

        int N = sc.nextInt();

        //System.out.println((1L << N) - 1);
        System.out.println(BigInteger.TWO.pow(N).subtract(BigInteger.ONE));

        if (N <= 20) {
            hanoiTower(N, 1, 3, 2);
        }

        sc.close();
    }
}
